src/Corax/Querying/Matches/TermMatch.cs

```diff
@@ -289,70 +289,84 @@ static QueryInspectionNode InspectFunc(ref TermMatch term)
             };
         }
         
-        public static TermMatch YieldSet(Querying.IndexSearcher indexSearcher, ByteStringContext ctx, PostingList postingList, double termRatioToWholeCollection, bool isBoosting, bool useAccelerated = true)
+        public static TermMatch YieldSet(IndexSearcher indexSearcher, ByteStringContext ctx, PostingList postingList, double termRatioToWholeCollection, bool isBoosting, bool useAccelerated = true)
         {
             [SkipLocalsInit]
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            static int AndWithFunc<TBoostingMode>(ref TermMatch term, Span<long> buffer, int matches) where TBoostingMode : IBoostingMarker
+            static int AndWithFunc<TBoostingMode>(ref TermMatch term, Span<long> buffer, int matchesCount) where TBoostingMode : IBoostingMarker
             {
-                int matchedIdx = 0;
-
+                if (matchesCount == 0)
+                    return 0;
+                
+                int resultCursor = 0;
+                int bufferCursor = 0;
+                
                 var it = term._set;
+                ref long matchesBuffer = ref MemoryMarshal.GetReference(buffer);
 
-                ref long start = ref MemoryMarshal.GetReference(buffer);
-                if (it.Seek(start - 1) == false)
+                var matchesLowest = Unsafe.Add(ref matchesBuffer, 0);
+                var matchesHighest = Unsafe.Add(ref matchesBuffer, matchesCount - 1);
+                
+                if (it.Seek(matchesBuffer - 1) == false)
                     return 0;
 
-                Span<long> decodedMatches = stackalloc long[1024];
+                Span<long> postingListItems = stackalloc long[1024];
+                
+                //Maximum value we can match.
+                var maxValidValue = Unsafe.Add(ref matchesBuffer, matchesCount - 1) + 1;
+                var currentMatchesDocument = Unsafe.Add(ref matchesBuffer, bufferCursor);
 
-                long maxValidValue = buffer[matches - 1] + 1;
-                while (true)
+                while (resultCursor < matchesCount && it.Fill(postingListItems, out var read, maxValidValue) && read > 0)
                 {
-                    if (it.Fill(decodedMatches, out var read, maxValidValue) == false)
-                        return matchedIdx;
-
-                    for (int j = 0; j < read; j++)
+                    var isFirstItemInsideRange = EntryIdEncodings.DecodeAndDiscardFrequency(postingListItems[0]) >= matchesLowest
+                                                && EntryIdEncodings.DecodeAndDiscardFrequency(postingListItems[0]) <= matchesHighest;
```

Comment:
This is a performance sensitive code, avoid calling twice `EntryIdEncodings.DecodeAndDiscardFrequency(postingListItems[0])` to shrink the method.